### PR TITLE
Support for app flavors in flutter tooling

### LINKED ISSUE #11547: Flutter CLI tools should understand build variants
https://github.com/flutter/flutter/pull/11354 added support for Android build variants in the Flutter Gradle plugin. But Flutter CLI tooling also need to know about build variants, so that `flutter run` and friends are able to do the correct thing.

### PR DESCRIPTION
**Pull request by mravn-google**
Fixes https://github.com/flutter/flutter/issues/11547.

`flutter build`, `flutter run`, and `flutter drive` now understand app "flavors" or "variants". This means that you can build different flavors of your app from the same source code, but with different configuration as defined by platform-specific build settings. Flavors are orthogonal to the existing `debug`/`profile`/`release` build modes:

```shell
flutter run --flavor free            # Launch the free app in debug mode
flutter run --flavor paid            # Launch the paid app in debug mode
flutter run --profile --flavor free  # Launch the free app in profile mode
flutter run --profile --flavor paid  # Launch the paid app in profile mode
flutter run --release --flavor free  # Launch the free app in release mode
flutter run --release --flavor paid  # Launch the paid app in release mode
```

Flavors at the Flutter CLI level map to [product flavors](https://developer.android.com/studio/build/build-variants.html#product-flavors) in Android projects (with Flutter build modes mapping to Android build types) and to [schemes](https://medium.com/practical-ios-development/some-practical-uses-for-xcode-build-schemes-and-build-configurations-swift-e50d15a1304f) in iOS projects (with Flutter build mode/flavor pairs mapping to Xcode build configurations).

The Android Gradle build tasks corresponding to the above would be
```
assembleFreeDebug
assemblePaidDebug
assembleFreeProfile
assemblePaidProfile
assembleFreeRelease
assemblePaidRelease
```
Android makes the flavor available to Java code as `BuildConfig.FLAVOR`. It can be communicated to the Dart side using platform channels (a convenient plugin for handling this could be built).

For Xcode, one can define build configurations according to a custom naming scheme, e.g.
```
Debug-Free
Debug-Paid
Release-Free
Release-Paid
```
---and then map each build configuration to one of two custom schemes named `Free` and `Paid`. Each build configuration might include a user-defined Build Setting entry, such as `FLAVOR` with values `free` or `paid`. This value can be exposed to ObjC by adding a `Flavor: $(FLAVOR)` key/value pair to Info.plist and then accessed using `[[NSBundle mainBundle].infoDictionary valueForKey:@"Flavor"]`. Flutter tooling does not care about how you set this up, it merely builds the Xcode project specifying the relevant build configuration and scheme.

A sample project using this setup is included for integration testing in `dev/integration_tests/flavors`. More documentation will be provided in a separate PR.

### CODE REVIEW

**Review by jakobr-google — APPROVED** [review]
LGTM

File: packages/flutter_tools/lib/src/android/gradle.dart [hunk 1]
```diff
@@ -299,3 +309,83 @@ Map<String, String> get _gradleEnv {
...
+      buildTypes.addAll(variants);
+    return new GradleProject(
+      buildTypes.toList(),
+      productFlavors.toList(),
+      fs.path.join(buildDir, 'outputs', 'apk'),
+    );
+  }
+
+  final List<String> buildTypes;
+  final List<String> productFlavors;
```
Comment (mravn-google) [inline packages/flutter_tools/lib/src/android/gradle.dart]: Yes. Lists are easier to create (for testing, defaults); there is no Set literal yet. And the expected length of these lists (2-4) is small enough that they are likely to outperform sets, even for `contains`.

**Comment by mit-mit** [comment]
Test run went fine; this is very nice!

**Comment by eseidelGoogle** [comment]
FYI @tvolkert @devoncarew

File: dev/devicelab/manifest.yaml [hunk 2]
```diff
@@ -59,6 +59,12 @@ tasks:
     stage: devicelab
     required_agent_capabilities: ["has-android-device"]
 
+  flavors_test:
+    description: >
+      Checks that flavored builds work on Android.
+    stage: devicelab
+    required_agent_capabilities: ["has-android-device"]
```
Comment (Hixie) [inline dev/devicelab/manifest.yaml]: please always add these as flaky

File: dev/integration_tests/flavors/android/gradle/wrapper/gradle-wrapper.properties [hunk 3]
```diff
@@ -0,0 +1,6 @@
+#Fri Jun 23 08:50:38 CEST 2017
```
Comment (Hixie) [inline dev/integration_tests/flavors/android/gradle/wrapper/gradle-wrapper.properties]: woah, you can't check in `gradle-wrapper.jar`

File: dev/integration_tests/flavors/android/gradlew [hunk 4]
```diff
@@ -0,0 +1,160 @@
+#!/usr/bin/env bash
```
Comment (Hixie) [inline dev/integration_tests/flavors/android/gradlew]: pretty sure this file shouldn't be getting checked in either

File: dev/integration_tests/flavors/android/gradlew.bat [hunk 5]
```diff
@@ -0,0 +1,90 @@
+@if "%DEBUG%" == "" @echo off
```
Comment (Hixie) [inline dev/integration_tests/flavors/android/gradlew.bat]: or this one

File: packages/flutter_tools/lib/src/build_info.dart [hunk 6]
```diff
@@ -8,10 +8,30 @@ import 'base/platform.dart';
...
+  /// If not null, the Gradle build task will be `assembleFlavorMode` (e.g.
+  /// `assemblePaidRelease`), and the Xcode build configuration will be
+  /// Mode-Flavor (e.g. Release-Paid).
+  final String flavor;
+
+  static const BuildInfo debug = const BuildInfo(BuildMode.debug, null);
+  static const BuildInfo profile = const BuildInfo(BuildMode.profile, null);
+  static const BuildInfo release = const BuildInfo(BuildMode.release, null);
+
+  const BuildInfo(this.mode, this.flavor);
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/build_info.dart]: constructor first please

File: packages/flutter_tools/lib/src/build_info.dart [hunk 7]
```diff
@@ -8,10 +8,30 @@ import 'base/platform.dart';
...
+
+  bool get isDebug => mode == BuildMode.debug;
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/build_info.dart]: these three should document that they are mutually exclusive

File: packages/flutter_tools/lib/src/commands/build_apk.dart [hunk 8]
```diff
@@ -35,6 +35,7 @@ class BuildApkCommand extends BuildSubCommand {
   BuildApkCommand() {
     usesTargetOption();
     addBuildModeFlags();
+    usesFlavorOption();
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/build_apk.dart]: isn't this just a build mode flag, essentially?

Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/build_apk.dart]: i guess not, never mind

File: packages/flutter_tools/lib/src/commands/build_ios.dart [hunk 9]
```diff
@@ -15,6 +15,7 @@ import 'build.dart';
 class BuildIOSCommand extends BuildSubCommand {
   BuildIOSCommand() {
     usesTargetOption();
+    usesFlavorOption();
     usesPubOption();
     argParser.addFlag('debug',
       negatable: false,
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/commands/build_ios.dart]: seems weird that this doesn't use addBuildModeFlags but that's not part of your patch

File: packages/flutter_tools/lib/src/ios/mac.dart [hunk 10]
```diff
@@ -356,7 +389,9 @@ Future<Null> diagnoseXcodeBuildFailure(XcodeBuildResult result, BuildableIOSApp
     printError(noDevelopmentTeamInstruction, emphasis: true);
     return;
   }
-  if (app.id?.contains('com.yourcompany') ?? false) {
+  if (result.xcodeBuildExecution != null &&
+      result.xcodeBuildExecution.buildForPhysicalDevice &&
+      app.id?.contains('com.yourcompany') ?? false) {
```
Comment (Hixie) [inline packages/flutter_tools/lib/src/ios/mac.dart]: we should really change this to com.example

File: dev/devicelab/manifest.yaml (hunk 2 above)
Comment (mravn-google) [inline dev/devicelab/manifest.yaml]: Done.

File: dev/integration_tests/flavors/android/gradle/wrapper/gradle-wrapper.properties (hunk 3 above)
Comment (mravn-google) [inline dev/integration_tests/flavors/android/gradle/wrapper/gradle-wrapper.properties]: Done.

File: dev/integration_tests/flavors/android/gradlew.bat (hunk 5 above)
Comment (mravn-google) [inline dev/integration_tests/flavors/android/gradlew.bat]: Done.

File: dev/integration_tests/flavors/android/gradlew (hunk 4 above)
Comment (mravn-google) [inline dev/integration_tests/flavors/android/gradlew]: Done.

File: packages/flutter_tools/lib/src/build_info.dart (hunk 6 above)
Comment (mravn-google) [inline packages/flutter_tools/lib/src/build_info.dart]: Done.

File: packages/flutter_tools/lib/src/build_info.dart (hunk 7 above)
Comment (mravn-google) [inline packages/flutter_tools/lib/src/build_info.dart]: Done.

File: packages/flutter_tools/lib/src/ios/mac.dart (hunk 10 above)
Comment (mravn-google) [inline packages/flutter_tools/lib/src/ios/mac.dart]: https://github.com/flutter/flutter/issues/11701

**Review by szakarias — APPROVED** [review]
LGTM

**Pull request merged**